server/service/async/async_scheduled_query_stats.go

```diff
@@ -0,0 +1,217 @@
+package async
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/config"
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+const (
+	scheduledQueryStatsHostQueriesKey = "scheduled_query_stats:{%d}" // the hash of scheduled query stats for a given host
+	scheduledQueryStatsHostIDsKey     = "scheduled_query_stats:active_host_ids"
+	scheduledQueryStatsKeyMinTTL      = 7 * 24 * time.Hour // 1 week
+)
+
+// RecordScheduledQueryStats records the scheduled query stats for a given host.
+func (t *Task) RecordScheduledQueryStats(ctx context.Context, hostID uint, stats []fleet.PackStats, ts time.Time) error {
+	cfg := t.taskConfigs[config.AsyncTaskScheduledQueryStats]
+	if !cfg.Enabled {
+		return t.datastore.SaveHostPackStats(ctx, hostID, stats)
+	}
+
+	// set an expiration on the  key, ensuring that if async processing is
+	// disabled, the key (eventually) does not use any redis space. Ensure that
+	// TTL is reasonably big to avoid deleting information that hasn't been
+	// collected yet - 1 week or 10 * the collector interval, whichever is
+	// biggest.
+	ttl := scheduledQueryStatsKeyMinTTL
+	if maxTTL := 10 * cfg.CollectInterval; maxTTL > ttl {
+		ttl = maxTTL
+	}
+
+	// the redis key is a hash where the field names are the
+	// packname\x00schedqueryname and the values are the json-marshaled
+	// fleet.ScheduledQueryStats struct. Using this structure ensures the fields
+	// are always overridden with the latest stats for the pack/query tuple,
+	// which mirrors the behavior of the mysql table (we only keep the latest
+	// stats).
+
+	// keys and arguments passed to the script are:
+	// KEYS[1]: hash key (scheduledQueryStatsHostQueriesKey)
+	// ARGV[1]: ttl for the key
+	// ARGV[2:]: arguments for HSET
+	script := redigo.NewScript(1, `
+    redis.call('HSET', KEYS[1], unpack(ARGV, 2))
+    return redis.call('EXPIRE', KEYS[1], ARGV[1])
+  `)
+
+	// build the arguments
+	key := fmt.Sprintf(scheduledQueryStatsHostQueriesKey, hostID)
+	args := redigo.Args{key, int(ttl.Seconds())}
+	for _, ps := range stats {
+		for _, qs := range ps.QueryStats {
+			field := fmt.Sprintf("%s\x00%s", ps.PackName, qs.ScheduledQueryName)
+			jsonStat, err := json.Marshal(qs)
+			if err != nil {
+				return ctxerr.Wrap(ctx, err, "marshal scheduled query stats")
+			}
+			args = args.Add(field, string(jsonStat))
+		}
+	}
+
+	if len(args) > 2 {
```

Comment:
When would `len(args)` be `<= 2`?

Reply:
if there aren't any items to save (args[0] is the key, args[1] the TTL, the rest are pairs of `(field,jsonStat)` items to store)

Reply:
Yeah the check is for what Roberto mentioned. I still call `storePurgeActiveHostID` as it may purge obsolete entries, so there's not much harm in doing this even if there were no stats to save for that host. In any case, I believe that's an unlikely event (to get called here with no stats).